database/schema/controller.go

```diff
@@ -17,28 +17,46 @@ CREATE UNIQUE INDEX idx_lease_type_type
 ON lease_type (type);
 
 INSERT INTO lease_type VALUES
-    (0, 'controller'), 
-    (1, 'model' ), 
-    (2, 'application');
+    (0, 'controller'),  -- The controller running singular controller workers.
+    (1, 'model' ),      -- The controller running singular workers for a model.
+    (2, 'application'); -- The unit that holds leadership for an application.
 
-CREATE TABLE IF NOT EXISTS lease (
+CREATE TABLE lease (
     uuid            TEXT PRIMARY KEY,
     lease_type_id   INT NOT NULL,
+    model_uuid      TEXT,
     name            TEXT,
     holder          TEXT,
     start           TIMESTAMP,
     expiry          TIMESTAMP,
-    pinned          BOOLEAN,
     CONSTRAINT      fk_lease_lease_type
         FOREIGN KEY (lease_type_id)
         REFERENCES  lease_type(id)
 );
 
-CREATE UNIQUE INDEX idx_lease_type_name
-ON lease (lease_type_id, name);
+CREATE UNIQUE INDEX idx_lease_model_type_name
+ON lease (model_uuid, lease_type_id, name);
 
 CREATE INDEX idx_lease_expiry
-ON lease (expiry);`[1:]
+ON lease (expiry);
+
+CREATE TABLE lease_pin (
+    -- The presence of entries in this table for a particular lease_uuid
+    -- implies that the lease in question is pinned and cannot expire.
+    uuid       TEXT PRIMARY KEY,
+    lease_uuid TEXT,
+    entity_id  TEXT,
+    CONSTRAINT      fk_lease_pin_lease
+        FOREIGN KEY (lease_uuid)
+        REFERENCES  lease(uuid)
+);
+
+CREATE UNIQUE INDEX idx_lease_pin_lease_entity
+ON lease_pin (lease_uuid, entity_id);
```

Comment:
why is the UNIQUE index on the tuple (lease_uuid, entity_id), certainly we can only have 1 pin per lease_uuid. Otherwise you would end up pinning 3 entities as the leader for a given lease. That also means you only need the one index rather than 2.

Reply:
The entity is the _requirer_ of the pin behaviour, not the holder to which it is pinned.